cli/cmd/install.go

```diff
@@ -4,26 +4,89 @@ package cmd
 )
 
+var installer = kubernetes.Installer{
+	Deployments: []kubernetes.Deployment{
+		&deployments.Traefik{},
+		&deployments.Quarks{},
+		&deployments.Gitea{},
+	},
+}
+
+func RegisterInstall(rootCmd *cobra.Command) {
+	installCmd := &cobra.Command{
+		Use:   "install",
+		Short: "install Carrier in your configured kubernetes cluster",
+		Long:  `install Carrier PaaS in your configured kubernetes cluster`,
+		Run:   Install,
+	}
+	installCmd.Flags().BoolP("verbose", "v", true, "Wether to print logs to stdout")
+
+	installer.GatherNeededOptions()
+	for _, opt := range installer.NeededOptions {
+		// Translate option name
+		flagName := strings.ReplaceAll(opt.Name, "_", "-")
+
+		// Declare option's flag, type-dependent
+		switch opt.Type {
+		case kubernetes.BooleanType:
+			installCmd.Flags().Bool(flagName, opt.Default.(bool), opt.Description)
+		case kubernetes.StringType:
+			installCmd.Flags().String(flagName, opt.Default.(string), opt.Description)
+		case kubernetes.IntType:
+			installCmd.Flags().Int(flagName, opt.Default.(int), opt.Description)
+		}
+	}
+
+	rootCmd.AddCommand(installCmd)
+}
+
 // Install command installs carrier on a configured cluster
 func Install(cmd *cobra.Command, args []string) {
 	fmt.Println("Carrier installing...")
-	installer := kubernetes.Installer{
-		Deployments: []kubernetes.Deployment{
-			&deployments.Traefik{},
-			&deployments.Quarks{},
-			&deployments.Gitea{},
-		},
-	}
-	installer.GatherNeededOptions()
-	installer.PopulateNeededOptions(nil)
+
 	cluster, err := kubernetes.NewCluster(os.Getenv("KUBECONFIG"))
 	ExitfIfError(err, "Couldn't get the cluster, check your config")
+
+	// Hack? Override static default for system domain with a
+	// function which queries the cluster for the necessary
+	// data. If that data could not be found the system will fall
+	// back to cli option and/or interactive entry by the user.
+	//
+	// NOTE: This is function is set here and not in the gitea
+	// definition because the function has to have access to the
+	// cluster in question, and that information is only available
+	// now, not at deployment declaration time.
+
+	domain, err := installer.NeededOptions.GetOpt("system_domain", "")
+	if err != nil {
+		ExitfIfError(err, "Couldn't install carrier")
+	}
+	domain.DynDefault = func(o *kubernetes.InstallationOption) error {
+		ips := cluster.GetPlatform().ExternalIPs()
+		if len(ips) > 0 {
+			domain := fmt.Sprintf("%s.nip.io", ips[0])
+			o.Value = domain
+			o.Valid = true
+		}
+		// else leave invalid, to be handled by cli option
+		// reader or interactive entry
+		return nil
+	}
+
+	installer.PopulateNeededOptions(kubernetes.NewDefaultOptionsReader())
+	installer.PopulateNeededOptions(kubernetes.NewCLIOptionsReader(cmd))
+	installer.PopulateNeededOptions(kubernetes.NewInteractiveOptionsReader(os.Stdout, os.Stdin))
+	// At last, check that all options are valid.
+	fmt.Println("Configuration...")
+	installer.PopulateNeededOptions(kubernetes.NewCheckOptionsReader())
```

Comment:
It is not truly "Validation" in the sense of "Is the value syntactically valid, etc". Only a check that the value got set by one of the preceding phases, and that nothing slipped through our nets.

Proper value validation I would put into `Validate()` method of the option, and have it called by each of the Readers, before they enter the user value (cli option or interactive) into the structures. I.e. early check in each phase instead of late check after all phases.

Reply:
For interactive entry this would also loop the entry prompt, as you already do for boolean options. (Could maybe subsumed into a standard validator for booleans).

Reply:
I will put the display of the entered configuration into a separate method of the installer.

Reply:
Addressed with commit f1027e9040f094d29697aa5e7a16c88ccd9e55cc